internal/component/loki/enrich/enrich.go

```diff
@@ -0,0 +1,196 @@
+// Package enrich provides the loki.enrich component.
+package enrich
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/alloy/internal/component"
+	"github.com/grafana/alloy/internal/component/common/loki"
+	"github.com/grafana/alloy/internal/component/discovery"
+	"github.com/grafana/alloy/internal/featuregate"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:      "loki.enrich",
+		Stability: featuregate.StabilityExperimental,
+		Args:      Arguments{},
+		Exports:   Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments configures the loki.enrich component.
+type Arguments struct {
+	// The targets to use for enrichment
+	Targets []discovery.Target `alloy:"targets,attr"`
+
+	// Which label from targets to use for matching (e.g. "hostname", "ip")
+	TargetMatchLabel string `alloy:"target_match_label,attr"`
+
+	// Which label from logs to match against (e.g. "hostname", "ip")
+	// If not specified, TargetMatchLabel will be used
+	LogsMatchLabel string `alloy:"logs_match_label,attr,optional"`
+
+	// List of target labels to copy to logs. If empty, all labels will be copied.
+	TargetLabels []string `alloy:"target_labels,attr,optional"`
+
+	// Where to forward logs after enrichment
+	ForwardTo []loki.LogsReceiver `alloy:"forward_to,attr"`
+}
+
+type Exports struct {
+	Receiver loki.LogsReceiver `alloy:"receiver,attr,optional"`
+}
+
+type Component struct {
+	opts    component.Options
+	args    Arguments
+	exports Exports
+
+	mut          sync.RWMutex
+	receiver     loki.LogsReceiver
+	targetsCache map[string]model.LabelSet
+	cacheMutex   sync.RWMutex
+}
+
+func New(opts component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:         opts,
+		args:         args,
+		targetsCache: make(map[string]model.LabelSet),
+		receiver:     loki.NewLogsReceiver(),
+	}
+
+	// Initialize the cache with provided targets
+	c.refreshCacheFromTargets(args.Targets)
```

Comment:
I think if the discovery component will just return an empty map if it hasn't discovered anything yet? Then `loki.enrich` would still not work for the first few seconds until the discovery starts working?

Reply:
hm, possibly. Is there any way to force discovery before component is initialized?

Reply:
is it acceptable?

Reply:
thanks for providing this example! I think I add a note for now.